clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -570,6 +604,21 @@ public void requestUpdate() {
             this.metadata.requestUpdate();
         }
 
+        /*
+         * Keep track of any nodes we've ever seen. Add current
+         * alive nodes to this tracking list. 
+         * @param nodes Current alive nodes
+         */
+        private void updateNodesEverSeen(List<Node> nodes) {
+            for (Node n : nodes) {
+                if (!nodesEverSeenById.containsKey(n.id())) {
```

Comment:
Yeah, this makes sense. You could always update the entry in `nodesEverSeenById`, make `nodesEverSeen` a `List<Integer>` containing node IDs, and only add a node ID to `nodesEverSeen` when `nodesEverSeenById` doesn't contain the ID as a key. Then looking up a random node requires using both, but it's cheap and shouldn't even be happening most of the time.